pyiron_contrib/atomistics/runner/utils.py

```diff
@@ -0,0 +1,125 @@
+"""Utility functions for use with the pyiron Hamiltonian of RuNNer.
+
+Attributes
+----------
+    RunnerTrainingContainer : TrainingContainer
+        Extension of the TrainingContainer class with a function to convert
+        to_ase.
+
+Reference
+---------
+    [RuNNer online documentation](https://theochem.gitlab.io/runner)
+"""
+
+from typing import List
+
+import numpy as np
+
+from ase.atoms import Atoms
+
+from pyiron_atomistics.atomistics.structure.atoms import pyiron_to_ase
+
+from runnerase.singlepoint import RunnerSinglePointCalculator
+
+from ..atomistics.job.trainingcontainer import TrainingContainer
+
+
+def container_to_ase(container: TrainingContainer) -> List[Atoms]:
+    """Convert a `TrainingContainer` into a list of ASE Atoms objects."""
+    structure_lst = []
+
+    arrays = container.to_dict()
+    arraynames = arrays.keys()
+
+    # Iterate over the structures by zipping the dictionary values.
+    for properties in zip(*arrays.values()):
+        zipped = dict(zip(arraynames, properties))
+
+        # Retrieve atomic positions, cell vectors, etc.
+        atoms = pyiron_to_ase(zipped['structure'])
+
+        # Attach properties to the Atoms object.
+        if 'charges' in zipped:
+            atoms.set_initial_charges(zipped['charges'])
+
+        if 'energy' in zipped:
+            energy = zipped['energy']
+        else:
+            energy = None
+
+        if 'forces' in zipped:
+            forces = zipped['forces']
+        else:
+            forces = None
+
+        if 'totalcharge' in zipped:
+            totalcharge = zipped['totalcharge']
+        else:
+            totalcharge = None
+
+        atoms.calc = RunnerSinglePointCalculator(
+            atoms=atoms,
+            energy=energy,
+            forces=forces,
+            totalcharge=totalcharge
+        )
+        structure_lst.append(atoms)
+
+    return structure_lst
+
+
+def ase_to_container(
+    structures: List[Atoms],
+    container: TrainingContainer
+) -> None:
+    """Add `structures` to `TrainingContainer`."""
+    for structure in structures:
+
+        properties = {}
+
+        # If the structure has a calculator attached to it, get energies,
+        # forces, etc.
+        if structure.calc is not None:
+            properties['energy'] = structure.get_potential_energy()
+            properties['forces'] = structure.get_forces()
+            properties['charges'] = structure.get_initial_charges()
+
+            if isinstance(structure.calc, RunnerSinglePointCalculator):
+                properties['totalcharge'] = structure.calc.get_property(
+                    'totalcharge'
+                )
+            else:
+                properties['totalcharge'] = np.sum(properties['charges'])
+
+        # StructureStorage needs a `spins` property, but not all ASE Atoms
+        # objects have that.
+        if not hasattr(structure, 'spins'):
+            structure.spins = None
```

Comment:
We only save and restore `spins` when available.  If that gave you and error, it's a bug we should fix in `StructureStorage`.

Reply:
I think it is because of line 202 f. in [pyiron_atomistics structurestorage.py](https://github.com/pyiron/pyiron_atomistics/blob/master/pyiron_atomistics/atomistics/structure/structurestorage.py):

```python
        if structure.spins is not None:
            arrays["spins"] = structure.spins
```

Maybe this could be solved with a simple `hasattr` check in the if-statement?

Reply:
Weird, will investigate next week.